chain/epoch-manager/src/lib.rs

```diff
@@ -1900,6 +1901,81 @@ impl EpochManager {
         }
     }
 
+    pub fn possible_epochs_of_height_around_tip(
+        &self,
+        tip: &Tip,
+        height: BlockHeight,
+    ) -> Result<Vec<EpochId>, EpochError> {
```

Comment:
I gave it a lot of thought and now IMO we can avoid this part of logic.
Let's say we always partially validate against current and next epoch. Previous epoch is not really needed.
Store predicted chunk producer' public key as part of key in the witnesses cache.
Once we receive block which matches the hash we can find what is the actual epoch id and CP' public key, and if state witness was signed incorrectly, we ban this CP.
This should be enough to restrict attempts to sign unexpected state witnesses on epoch boundaries to wipe out cache. WDYT?

Reply:
If we validate the witness against the current and next epoch, without looking if the next epoch is even possible, it would give a malicious chunk producer a much bigger ability to spam the cache with fake witnesses. This chunk producer would be able to take up a spot in the cache for every (shard, height) for which it's a producer in this or next epoch.

For 99% of the time the tip is in a position where `height + 5` is known to be in the same epoch, so there's only one  possible epoch and we're able to validate that the witness came exactly from the chunk producer from which it should. This really reduces the amount of damage that a single chunk producer can do - for 99% of the time they can only take up the spot that is theirs anyway, without affecting witnesses produced by other lawful chunk producers. The only time they can potentially take up a spot which doesn't belong to them is around epoch boundaries (5/43200 = 0.000115 of blocks). I really like that for 99% of blocks we have the guarantee that no one but the real chunk producer can take up a spot in the cache.

If we allow witnesses that would belong to a producer in the next epoch, that allows the malicious chunk producer to take up a spot for every block that it produces in the next epoch, which is much larger than 5/43200.
We can key by the signer public key to not eject the real chunk producer's witness, but it still takes up a spot in  the cache, and the cache is really small. In the current version the cache has only capacity of ~3 witnesses per shard, so by taking up a single spot the malicious chunk producer would reduce the capacity by 33%. With small items like chunk endorsements, there is no problem, we can easily store hundreds of them to mitigate fake entries, but this cache has to be small because the witnesses can be 40MB in size. Because of that I think that we should be as strict as possible with the validation. Maybe with single-shard tracking we would be able to increase the per-shard capacity, but for now there are only 3 spots :/

> Once we receive block which matches the hash we can find what is the actual epoch id and CP' public key, and if state witness was signed incorrectly, we ban this CP.

This might not work, as a malicious chunk producer can submit a witness that waits for a nonexistant block (`hash("1234")`). We could try validating the witnesses once we reach their height, but I'm worried that there might be some corner cases with forks where we could ban an innocent chunk producer. I don't fully understand how forks work, so I'm not sure, I would have to read up on it.

It's a tradeoff between complexity and security. I really liked that for 99% of blocks we are able to verify the exact chunk producer, so I went with `possible_epochs_of_height_around_tip`, but maybe we could afford to have some spam entries to reduce the complexity 🤔

Reply:
I think I finally understood why you proposed the per-chunk-producer cache first - in such case chunk producer would spoil only their life.
Makes sense, thank you for the thorough explanation!

Reply:
It's possible that it would be secure enough without `possible_epochs_of_height_around_tip`, but it's hard for me to say how much worse it would be. It sounds tempting, but I would have to prepare some estimations to convince myself that it's still ok to do. That sounds a lot of work (and complexity in the analysis), so I'm leaning towards leaving it as is :P